### DISCUSSION TITLE
Lists of options for form select lists

### DISCUSSION THREAD

**Discussion by mikevarela** [prisma/prisma#18110] [Q&A] [answered]
I've completed my data models and now in the frontend (NextJS / Postgresql), I'm attempting to fill some HTML select lists for a form I'm creating. The question revolves around getting the select list data options from a number of tables in the database. 

Let's say I have a user. Currently, they have some personal details that I'd like to store with the user themselves. The options might be

- gender
- sexual orientation
- race
- type

etc.

I've created models for each option, each with an ID field and a Value field. No relations exist in the DB for these options.

The User model provides field definitions for these options as Strings, this way I don't have to perform a number of lookups to search and grab a few users and their associated options.

However, my issue is in creating a 'New User' form. I need to first grab all the options before putting those into select lists in the HTML. That's a rather expensive series of Prisma calls. Ideally, each of these lists are backed with CRUD REST APIS's so a user could add or change the options down the line. 

I have two questions

1. what's the best way to query these efficiently? Currently, I'm doing a series of async/await calls, one after the other. It seems laborious and prone to errors.
2. The returned responses of each call result in an array of objects, but, I'm actually interested in an array of strings for each, that way I'm able to easily map these into select lists. Is there a way to grab the values of each object without performing a loop on each return?

And, forgive the barrage of questions, in terms of design, what's the best way to manage this type of workflow? This form is going to have several select lists and ideally, each is backed by a REST API with CRUD operations. I know I could create a large options object or ENUMS I suppose but that seems inflexible.

Thanks so much, loving Prisma!

**Comment by nurul3101** [comment, marked as answer]
Hey @mikevarela 👋 

For question 1, can you try using [Promise.all](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all) to execute multiple queries in parallel?

For question 2, you can use [Prisma Client Extension](https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions/) to transform the response before sending it to the client. But I believe you might need to use a for loop in this case as well.

Overall, the best approach will depend on the specific requirements of your application. If you expect to have a large number of select lists, each with a large number of options, then it may be more efficient to create separate API endpoints for each select list. However, if you only have a few select lists with a small number of options, then a single API endpoint may be simpler to manage.

**Comment by duron23** [comment]
1. Create an API end point to fetch the options data for each of your select list.
2. Separate each select list into its own client components. 
3. Use react-query hook to fetch the list options for each of the select options from api end point created in step 1.
4. Use this client components in your forms.

Advantage of this approach is
Reusability: These select options can be used anywhere you need similar select options.
Cache: React Query will take care of caching, hence less db operations.
Error Handling: you have better error handling.
Design patterns: Single responsibility principle.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
